CARDS-1400: New Questionnaire entry type: Question matrix

veronikaslc opened the pull request:
[CARDS-1400](https://phenotips.atlassian.net/browse/CARDS-1400): New Questionnaire entry type: Question matrix

**Testing**
* Run with the `--test` flag and test the `Question Matrix Test` questionnaire,
OR
* Build the `CARDS-1592_CARDS-1594` branch (PR #1030), which is developed on top of this branch, and includes the `SF12` questionnaire with several `QuestionMatrix` form entries.

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Draft data modeling commit

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Corrected data modeling

Commit:
[misc] Removed unused import

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Form rendering, Subject, Subject timeline, form to text adapter

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Questionnaire wizard

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Some bugfix

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: Form rendering bugfixes

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: some UI polishing

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
WIP: QuestionMatrix Commit editor

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Subject view rendering

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Removed QuestionMatrix data type

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Some more adjustments regarding removing QuestionMatrix data type

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Renamed component to QuestionMatrix

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Checking props existing before copying to the question

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Added missing props setting when option is copied to the new question

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -627,6 +627,10 @@ const questionnaireStyle = theme => ({
       paddingLeft: theme.spacing(4)
     },
+    tableCell: {
+      textAlign: "middle",
```
`middle` is not a supported value for `test-align`.
```suggestion
      textAlign: "center",
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -348,25 +361,116 @@ Question.propTypes = {
+												          label: "Condition",
+												          value : value?.operandA?.value.join(', ') + " " + value?.comparator + " " + value?.operandB?.value.join(', ')
+												        });
+												      });
```
The code is hard to read with this large indentation

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -28,16 +28,18 @@ import FormattedText from "../components/FormattedText.jsx";
 function Question (props) {
-  let { classes, children, questionDefinition, existingAnswer, isEdit, preventDefaultView, defaultDisplayFormatter } = props;
-  let { text, compact, description, disableInstructions } = { ...questionDefinition, ...props }
+  let { classes, children, questionDefinition, existingAnswer, answers, isEdit, preventDefaultView, defaultDisplayFormatter } = props;
+  let { text, label, compact, description, disableInstructions } = { ...questionDefinition, ...props }
+
+  let existingAnswers = answers || (existingAnswer ? Array.of(existingAnswer[1]["displayedValue"]).flat() : null);
```
Instead of adding all these props and making changes to the `Question` component, the `QuestionMatrix` component should pass modified objects as props so the existing `Question` can be used.

veronikaslc replied:
Not sure there is more optimal way to pass `answers  from `QuestionMatrix`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -0,0 +1,243 @@
+
+  return (
+    <Question
+      questionDefinition={sectionDefinition}
```
Create a clone of `sectionDefinition`, then replace the `label` field with `text` so that the `Question` component doesn't need to be modified.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -0,0 +1,243 @@
+  return (
+    <Question
+      questionDefinition={sectionDefinition}
+      answers={existingAnswers}
```
Instead of a new prop `answers`, re-organize `existingAnswers` into the format expected from `existingAnswer` in view mode.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -0,0 +1,243 @@
+            questionName={question[0]}
+            {...rest}
+          />
+      )}
```
Some pre-processing of `subquestions` would be useful here to have meaningful variable names, make the code easier to read and avoid repeating fragmments like `question[0]`, `question[1]["jcr:uuid"]`. This can be coupled with the reorganizing of `existingAnswers` suggested for line 165.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -52,7 +52,7 @@ import QuestionnairePreview from "./QuestionnairePreview";
 let _stripCardsNamespace = str => str.replaceAll(/^cards:/g, "");
 
-export const QUESTIONNAIRE_ITEM_NAMES = ENTRY_TYPES.map(type => _stripCardsNamespace(type));
+const QUESTIONNAIRE_ITEM_NAMES = QUESTION_TYPES.concat(SECTION_TYPES).concat(INFO_TYPES).map(type => _stripCardsNamespace(type));
```
Replacing `ENTRY_TYPES` with `QUESTION_TYPES.concat(SECTION_TYPES).concat(INFO_TYPES)` seems useless here, please revert.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -647,6 +647,9 @@ export function displayQuestion(entryDefinition, data, key, classes) {
   const isHidden = (entryDefinition.displayMode == "hidden");
+  if (isHidden) {
+    return null;
+  }
```
Why was this change necessary, how does it relate to the implementation of QuestionMatrix?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectTimeline.jsx:
```diff
@@ -231,31 +231,30 @@ function SubjectTimeline(props) {
+	          // if a previous date answer exists and hasn't met the followup question limit.
+	          currentSectionData[currentSectionData.length - 1].followup.push(displayQuestion(entryDefinition, data, key, classes));
+	      }
+	    }
```
Unless I'm missing something, all the changes in this file appear to be whitespace changes in a file otherwise not affected in any way by the introduction of QuestionMatrix. Please revert and make a separate PR for formatting.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionMatrixEditor.java:
```diff
@@ -0,0 +1,258 @@
+        if (source.hasProperty("notApplicable")) {
+            Boolean notApplicable = source.getProperty("notApplicable").getValue(Type.BOOLEAN);
+            newNode.setProperty("notApplicable", notApplicable, Type.BOOLEAN);
+        }
```
Is there a built-in copy or clone method that can be used instead?

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionMatrixEditor.java:
```diff
@@ -0,0 +1,258 @@
+                        childNode.setProperty(propName, after.getValue(Type.STRING), Type.STRING);
+                    } else {
+                        childNode.setProperty(propName, after.getValue(Type.LONG), Type.LONG);
+                    }
```
`after` should know its data type so that we don't need to hard-code per property.

Also set the `displayMode` to `list` here to make filters work.

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed answer instructions alignment;
Do not show questions without answers in form view mode

marta- reviewed (commented):
Summary of outstanding changes, details inline:
* Commit editor: improve AnswerOption cloning, refactor property transfer from section to sub-questions, set `displayMode=list` on sub-questions
* Refactor QuestionMatrix to avoid changes to Question.jsx
* Some unnecessary changes need to be reverted

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Addressed codereview comments:
set displayMode=list on sub-questions
reverted unnecessary changes
refactored QuestionMatrix to avoid changes to Question.jsx

veronikaslc commented:
TODO:
- fix Quick Search NPE while searching for one of the AnswerOption values associated with a Question matrix results
- fix markdown description rendering for Questions related to matrix section
- investigate the node copy method `org.apache.sling.api.resource.ResourceResolver.copy(String, String)`

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed show section in view mode if mandatory

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed Quick Search NPE while searching for one of the AnswerOption values associated with a Question matrix results

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Moved highlight logic to the Question to avoid code duplication

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Fields.jsx:
```diff
@@ -91,17 +92,21 @@ let Fields = (props) => {
   JSON: PropTypes.object.isRequired,
-  edit: PropTypes.bool.isRequired
+  edit: PropTypes.bool.isRequired,
+  isMatrix: PropTypes.bool,
```
why did this need to be added?

veronikaslc replied:
To be used in `AnswerOptions.jsx`

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionMatrixEditor.java:
```diff
@@ -0,0 +1,259 @@
+            for (String childNodeName : this.currentNodeBuilder.getChildNodeNames()) {
+                NodeBuilder childNode = this.currentNodeBuilder.getChildNode(childNodeName);
+                if (isQuestion(childNode)) {
+                    if (StringUtils.equals(propName, "dataType")) {
+                        childNode.setProperty(propName, after.getValue(Type.STRING));
+                    } else {
+                        childNode.setProperty(propName, after.getValue(Type.LONG));
+                    }
```
This still needs to be refactored to get the value type from `after` and remove the `if ... else`.

marta- commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -308,11 +308,26 @@
   // Can this section be repeated?
   - recurrent (boolean) = false autocreated
 
+  - displayMode (string)
```
```suggestion
  // Section display mode, one of default, header, footer. matrix
  - displayMode (string) = 'default' autocreated
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionnaireQuickSearchEngine.java:
```diff
@@ -52,6 +52,7 @@
     }
 
     @Override
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity"})
```
Could we refactor by pulling lines 73-106 into a separate function `private void addResult(final Resource res, final List<JsonObject> output)` to avoid this warning?

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Addressed codereview comments: optimisation and refactoring

sashaandjic commented:
QUESTION: Should changing section's `Display mode: default` to `Matrix` of the existing `Proms` EQ-5D form work?
I made all necessary changes, however I cannot get the proper appearance of the matrix when in the edit mode?
Please note, making a brand new section with the `Display mode: Matrix` works as expected.
![image](https://user-images.githubusercontent.com/23243244/145236813-5b4e10f1-597a-44a7-9981-cbc90c020168.png)

marta- commented:
> QUESTION: Should changing section's `Display mode: default` to `Matrix` of the existing `Proms` EQ-5D form work? I made all necessary changes, however I cannot get the proper appearance of the matrix when in the edit mode? Please note, making a brand new section with the `Display mode: Matrix` works as expected.

Simply changing the display mode to matrix without specifying answer options **in the section** will not work. Once you change the display mode to matrix, the inner questions lose their own settings (except text and description) and inherit the properties specified in the matrix section. The result you show in the screenshot is expected based on that workflow. You must specify `dataType`, `min/maxAnswers` and AnswerOptions in the section for the whole thing to work properly.

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Added test questionnaire

marta- commented:
Added test questionnaire in 4d8e7b7,

Functionality-wise, there seem to be 2 more issues:
* Possible regression: values instead of labels are displayed in view mode. To replicate, fill out the newly added test questionnaire, save and view.
* CARDS-1006: _When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts_ has not been extended to QuestionMatrix. The questions in the matrix that don't have an answer are not displayed in view modem even if they are mandatory. To replicate, instantiate the QuetionMatrixTest and leave a couple of questions unanswered before clicking save&view.

sashaandjic commented:
I answered 2 questions out of many:
![image](https://user-images.githubusercontent.com/23243244/145438314-9487f780-895d-46a7-a00b-a35766ffdafa.png)

marta- commented:
Additional bugs:

1/ Illustrated in the 2nd screenshot in @sashaandjic's comment:
* create a form for a questionnaire that includes a question matrix,
* fill out some answers,
* click save & view,
* then click edit **-> the radio buttons corresponding to the provided answers do not appear checked**

2/ Save issues:
* create a form for a questionnaire that includes a question matrix,
* fill out some answers,
* click SAVE,
* then click save & view **-> the user is NOT directed to view mode**
* inspect network tab -> **the 2nd save call response is** `409 repository state conflicting with request`

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Bug fix: for question matrices with numeric datatype, the selection appears to be missing on re-edit

marta- commented:
> * create a form for a questionnaire that includes a question matrix,
> * fill out some answers,
> * click save & view,
> * then click edit **-> the radio buttons corresponding to the provided answers do not appear checked**

Fixed in 0857740

marta- commented:
Bug: placing a question matrix inside a recurring section causes a server error when attempting to save the second instance of the recurring section.

```
status.code: 500
status.message: "java.lang.IllegalStateException: Duplicate key [Ljava.lang.Object;@5768d04"
title: "Error while processing /Forms/f3194ea5-e440-408f-a933-81695b98dd41"
```

sashaandjic commented:
>

sashaandjic commented:
> > * create a form for a questionnaire that includes a question matrix,
> > * fill out some answers,
> > * click save & view,
> > * then click edit **-> the radio buttons corresponding to the provided answers do not appear checked**
> 
> Fixed in [0857740](https://github.com/data-team-uhn/cards/commit/085774037fb4fe8c16799e8dc3b800f8f5c88db2)

rested

sashaandjic commented:
> Bug: placing a question matrix inside a recurring section causes a server error when attempting to save the second instance of the recurring section.
> 
> ```
> status.code: 500
> status.message: "java.lang.IllegalStateException: Duplicate key [Ljava.lang.Object;@5768d04"
> title: "Error while processing /Forms/f3194ea5-e440-408f-a933-81695b98dd41"
> ```
Confirmed
Clicking on SAVE button brings the login dialog
![image](https://user-images.githubusercontent.com/23243244/145497033-ccb9d233-cb1a-4d6b-bd7f-9e76eda49c69.png)

marta- commented:
> * CARDS-1006: _When in a view mode of an invalid/incomplete form, show all the offending fields marked with proper message in red fonts_ has not been extended to QuestionMatrix. The questions in the matrix that don't have an answer are not displayed in view modem even if they are mandatory. To replicate, instantiate the QuetionMatrixTest and leave a couple of questions unanswered before clicking save&view.

The frontend code appears correct, however the `statusFlags` are always `[]` in the `.deep.json` for the matrix section and its children.

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Copy all matrix sextion props to new Question child node

veronikaslc commented:
> The questions in the matrix that don't have an answer are not displayed in view modem even if they are mandatory. 

Fixed in 55a74db

marta- commented:
> > The questions in the matrix that don't have an answer are not displayed in view modem even if they are mandatory.
> 
> Fixed in [55a74db](https://github.com/data-team-uhn/cards/commit/55a74dbbe995c2a33d397d889e354f813cae8321)

Question matrices created using the questionnaire designer are now displayed as expected.

The test questionnaire still doesn't display incomplete questions in view mode.

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Corrected comment

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Renamed gridClasses to cardClasses

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Code cleanup

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed 409 repository state conflicting with request when saving newly created Form first time with matrix quetion with an answer by clicking save and then save&view

veronikaslc commented:
TODO:
- 500 java.lang.IllegalStateException: Duplicate key ... when saving matirx quetion in second instance in recurrent section
- Copy Matrix Section Answer options and properties to the questions on the questionnaire import

sdumitriu committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed AnswerOption copying on initial content import

sdumitriu committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed wrong property type for jcr:primaryType on copies answers

marta- commented:
409 error and importing the test form both appear to be fixed based on light testing.
500 error is the only outstanding functionality issue at the moment.

marta- commented:
The spacing is quite unbalanced when there are unanswered subquestions:
![image](https://user-images.githubusercontent.com/651980/145896192-834934b2-9642-43ae-93bb-29e401727624.png)

Can we display answer instructions only once under the main title and color unanswered subquestions in red instead?
![image](https://user-images.githubusercontent.com/651980/145896822-b5f33a8c-aee8-4047-9496-7128d19e3177.png)

sashaandjic commented:
I would also replace the answered  options with smaller and bolded fonts.
![image](https://user-images.githubusercontent.com/23243244/145898228-9e727cad-30a4-413c-a839-acb483cb4763.png)

Same size as when in edit mode:
![image](https://user-images.githubusercontent.com/23243244/145898279-d5283e21-7ea0-48b3-9281-53a6f2fc06f0.png)

sashaandjic commented:
> Can we display answer instructions only once under the main title and color unanswered subquestions in red instead?

That would be awesome.

marta- commented:
> EDITED: I would also replace the answered options with smaller and bolded fonts.

That will likely make everything look a bit too bold IMO, and hard to distinguish between question and answer. Instead, for a more balanced look, I suggest using `subtitle1` typography instead of the current `subtitle2` for subquestions (which will have the same font size as the answers and will be 1 level less "bold" than the current).

> How about getting answers a bit closer to the question?

+1. Can we have a table instead of grids in view mode, with auto width? This way the contents will all be aligned while the answers will be naturally close enough to their questions.

marta- commented:
Summary of UI changes:
* Revert current changes to `AnswerInstructions.jsx`, as they will become unnecessary
* Display answer instructions only once under the main question text (exactly as it is currently done in all question types). If at least one of the subquestions does not satisfy the min/max answers requirements, the instructions will be displayed in red.
* Any subquestion that does not satisfy the min/max answers requirement will be displayed in `error` color
* Use `subtitle1` instead of `subtitle2` for subquestions
* In view mode, use a table that **does not** stretch to full width to display questions and their answers (or any other layout method that brings answers close to their subquestions while still keeping everything vertically aligned).

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Minor UI corrections

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -169,35 +169,37 @@ let QuestionMatrix = (props) => {
   }
 
-  let existingAnswerMock = [sectionAnswerPath, {displayedValue: existingAnswers}];
+  let existingAnswerMock = [sectionAnswerPath, {displayedValue: existingAnswers, statusFlags: existingSectionAnswer[1].statusFlags}];
```
This change will cause a js error when `existingSectionAnswer` is `undefined`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -169,35 +169,37 @@ let QuestionMatrix = (props) => {
-  let existingAnswerMock = [sectionAnswerPath, {displayedValue: existingAnswers}];
+  let existingAnswerMock = [sectionAnswerPath, {displayedValue: existingAnswers, statusFlags: existingSectionAnswer[1].statusFlags}];
   let newQuestionDefinition = { ...sectionDefinition, text: sectionDefinition.label};
+  let currentAnswers = Math.floor(Object.keys(selection).reduce((sum, item) => {return sum + selection[item].length;}, 0) / subquestions.length);
```
If, on the test questionnaire, we set minAnswers=1 and maxAnswers=2, we can tick 2 checkboxes for the first 4 questions to make the answer instructions turn from red to gray suggesting all answers are valid, which is not true. `currentAnswers` should be calculated as the min, not the average.

veronikaslc replied:
Oh, right, how did I now think of it!

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionMatrix.jsx:
```diff
@@ -169,35 +169,37 @@ let QuestionMatrix = (props) => {
+              </TableCell>
+            </TableRow>
+          </TableBody>
+        </Table>
```
This will generate separate tables for each subquestion, each within a list item, meaning the displayed answers will not be vertically aligned. Instead of `defaultDisplayFormatter`, use the `preventDefaultView` prop and generate a single table for the entire subsection list.

Commit:
CARDS-1400: New Questionnaire entry type: Question matrix
Minor UI corrections

veronikaslc commented:
TODO:
- 500 java.lang.IllegalStateException: Duplicate key ... when saving matrix quetion in second instance in recurrent section

sashaandjic commented:
Current look:
![image](https://user-images.githubusercontent.com/23243244/146282875-30eb7689-9f99-453f-9d04-17ac7c4fa70d.png)

Even though the answers appear to be in smaller fonts, the look of the form in view mode is OK.

Commit:
CARDS-1070: Changing the question variable name erases the previously entered AnswerOptions

veronikaslc commented:
Used `subtitle1` typography as was recommended

sdumitriu reviewed (commented):
Two (related) bugs:
- create a new questionnaire
- add: section
- **do not** set a section identifier
- display mode: matrix
- add 3 answer options
- Save
- the section doesn't appear there (bug 1a)
- refresh -> 4 sections show up, an empty matrix section and 3 empty normal sections (bug 1b)

And:
- create a new questionnaire
- add a regular section section
- start adding a subsection in it
- **do not** set a section identifier
- display mode: matrix
- add 3 answer options
- Save
- red error displayed instead of the subsection (bug 2a)
- trying to add a subsection again doesn't work (bug 2b)

sashaandjic commented:
I can no longer reproduce problem with placing a question matrix inside a recurring section (`status.code: 500`).
Confiming 2 problems that Sergiu reported. Why don't we enforce ID for sections and question to always be unique and not blank?
 
<img width="815" alt="image" src="https://user-images.githubusercontent.com/23243244/146629181-4c6ae6bf-2016-471f-8166-ddb9fdd4bd80.png">
<img width="251" alt="image" src="https://user-images.githubusercontent.com/23243244/146629185-8093a820-d60b-46e4-9eb7-5c9bc5742db8.png">
<img width="810" alt="image" src="https://user-images.githubusercontent.com/23243244/146629191-fc5f29d2-3cd4-4701-b60e-a4de9af3f193.png">

veronikaslc commented:
Those issues are from different tasks and addressed  in CARDS-1070 and CARDS-692 PRs.

sashaandjic commented:
> Those issues are from different tasks and addressed in CARDS-1070 and CARDS-692 PRs.

Ok, then I will mark this PR as tested.

sashaandjic reviewed (approved):
tested ( the issues reported by Sergiu are addressed in CARDS-1070 and CARDS-692 PRs)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -172,6 +143,54 @@ let displayInformation = (infoDefinition, key, classes, pageActive, isEdit) => {
+ * @param {Object} classes style classes
+ * @returns a React component that renders the matrix section
+ */
+let displayMatrix = (sectionDefinition, path, existingAnswer, key, classes, pageActive, isEdit, instanceId) => {
```
Not really introduced in this PR, since this was copied from above, but there seems to be some new parameters introduced but not documented.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -172,6 +143,54 @@ let displayInformation = (infoDefinition, key, classes, pageActive, isEdit) => {
+      && answer[1]["sling:resourceSuperType"] === "cards/Answer");
+
+  // Determine if the section is flagged as incomplete/invalid
+  // Note: bay be better explicitly check <statusFlags.includes('INCOMPLETE')> and same for INVALID
```
```suggestion
  // Note: may be better to explicitly check <statusFlags.includes('INCOMPLETE')>, and same for INVALID
```

sashaandjic reviewed (approved):
re-tested
Problem with recurring sections containing matrix is resolved.
<img width="733" alt="image" src="https://user-images.githubusercontent.com/23243244/147693150-52c7b616-6b42-4829-9aab-963947db93de.png">

veronikaslc commented:
Rebased on the latest dev.

marta- commented:
This PR needs to be retested after rebase on dev.

sashaandjic reviewed (approved):
re-tested after rebase

acrowthe committed:
CARDS-1400: New Questionnaire entry type: Question matrix
- Implement pageActive support

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -449,7 +449,7 @@ let Section = (props) => {
 
   if (data.displayMode == "matrix") {
-    return <QuestionMatrix {...props} />
+    return <QuestionMatrix {...props} pageActive={true} />
```
`pageActive={true}` is not necessary here. The `QuestionMatrix` component being used is not the one rendered in forms, but another with the same name (hence the confusion!) defined earlier in this file at line 364, used for rendering the QuestionMatrix definition in the questionnaire's edit mode.

acrowthe committed:
CARDS-1400: New Questionnaire entry type: Question matrix
- Remove unneeded pageActive from administration page

marta- committed:
[misc] Bug fix: `compact` always ends up being `undefined` in the `Question` component

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
* Responsive layout for QuestionMatrix: display entries similar to regular questions (vertival layout) on narrow screens to avoid some answers options becoming out of reach
* Styling improvements for view mode
* Remove unnecessary `label` prop introduced previously to the Question component
* Other minor code comments and code style improvements

marta- commented:
Notable changes in the latest push:
* Rebased on latest `dev`
* Bug fix: `compact` layout for questions now works - 4d92e49
* Made the QuestionMatrix responsive (i.e. on narrow screens the questions inside the matrix switch to vertical layout, looking more like regular questions, to ensure all answer options are visible) - d0efc5a

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Code style: removed unneeded imports, cleaned up indentation

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Allow N/A options to be placed both at the beginning and at the end of the list of answer options for a question matrix

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Refactored the Questionnaire Wizard support for QuestionMatrix to keep the code more generic rather than hardcoding logic for the `displayMode=matrix` case

marta- commented:
Notable changes in the latest push:

* Refactored the Questionnaire Wizard support for QuestionMatrix; there should be no visible changes to the user in commit 2a63612
* Enabled support for N/A options both at the beginning and at the end the option list in commit c839cef (before this commit, N/A could be added only at the end of the option list, per initial specs)
  * Note: imo it is ok that the default label of the end NA option is "None of the **above**" even though the default layout of a question matrix is horizontal; the label is easy to change at the time of designing the question matrix

marta- commented:
Notable changes in the latest push:

* Refactored the Questionnaire Wizard support for QuestionMatrix; there should be no visible changes to the user in commit 6c03415
* Enabled support for N/A options both at the beginning and at the end the option list in commit 829e890 (before this commit, N/A could be added only at the end of the option list, per initial specs)
  * Note: imo it is ok that the default label of the end NA option is "None of the **above**" even though the default layout of a question matrix is horizontal; the label is easy to change at the time of designing the question matrix

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Fixed regression introduced by previous refactoring commit: attemting to add a new item to a questionnaire fails with a white screen

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Small refactoring: make the way we look for `//CHILDREN` overrides more generic

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Small spacing adjustment for Question Matrix with horizontal layout

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -28,13 +29,49 @@ import FormattedText from "../components/FormattedText.jsx";
 function Question (props) {
-  let { classes, children, questionDefinition, existingAnswer, isEdit, pageActive, preventDefaultView, defaultDisplayFormatter } = props;
-  let { text, compact, description, disableInstructions } = { ...questionDefinition, ...props }
+  let { classes, children, questionDefinition, existingAnswer, answers, isEdit, pageActive, preventDefaultView, defaultDisplayFormatter } = props;
```
This looks like a leftover change.
```suggestion
  let { classes, children, questionDefinition, existingAnswer, isEdit, pageActive, preventDefaultView, defaultDisplayFormatter } = props;
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -436,3 +474,38 @@ let FieldsGrid = (props) => {
+      </Grid>
+    </Grid>
+  );
+}
```
```suggestion
}

```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -131,6 +131,10 @@ const questionnaireStyle = theme => ({
         margin: theme.spacing(-3,0,1),
         padding: theme.spacing(1, 0),
     },
+    answerMatrixInstructions: {
+        margin: theme.spacing(-0.5,0,1),
+        padding: theme.spacing(1, 0),
+    },
```
This doesn't appear to be used.

```suggestion
```

marta- committed:
[misc] Clean up leftover changes

sashaandjic reviewed (approved):
tested

sashaandjic reviewed (approved):
tested

acrowthe commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionMatrixEditor.java:
```diff
@@ -0,0 +1,259 @@
+            Boolean notApplicable = source.getProperty("notApplicable").getValue(Type.BOOLEAN);
+            newNode.setProperty("notApplicable", notApplicable, Type.BOOLEAN);
+        }
+    }
```
```suggestion
        if (source.hasProperty("noneOfTheAbove")) {
            Boolean noneOfTheAbove = source.getProperty("noneOfTheAbove").getValue(Type.BOOLEAN);
            newNode.setProperty("noneOfTheAbove", noneOfTheAbove, Type.BOOLEAN);
        }
    }
```
Should this be copied over like the `notApplicable` property? `noneOfTheAbove` options seem to work fine without this property being copied over, but I'm not sure if this being missed would cause issues in the future.

marta- replied:
Yes, good catch

marta- replied:
Actually, all works because when displaying forms with question matrices we read all the answer options info from the section and ignore anything specified in the question. Copying the options to the question itself is for filtering purposes only. So technically both the `notApplicable` property and the `noneOfTheAbove` property are unnecessary, but if we're going to copy one we'd better do the same for the other for consistency.

marta- replied:
Done in 5a652b9.

acrowthe reviewed (commented):
The questionnaire editor has the ability to add descriptions to question matrix answer, but they are not displayed on the form editor:
![image](https://user-images.githubusercontent.com/20056751/168623078-6086a73c-5d51-478b-9e78-169ac4c1f96d.png)
![image](https://user-images.githubusercontent.com/20056751/168623350-19854287-fce7-4366-b314-b7d4965d7874.png)

With multi select checkboxes, `none` and `notApplicable` checkboxes cannot be manually deselected:
https://user-images.githubusercontent.com/20056751/168624139-4102142e-d7da-467c-b7c3-e5bab3f7ebd7.mp4

marta- commented:
> The questionnaire editor has the ability to add descriptions to question matrix answer, but they are not displayed on the form editor:

Right, we display the question descriptions but not the answer descriptions. Will take care of it.

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix

Also copy the `noneOfTheAbove` property to the AnswerOption of the questions inside a question matrix

Co-authored-by: acrowthe <acrowthe3388@gmail.com>

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Show answer option descriptions in the question matrix

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Bug fix: when maxAnswers != 1, `notApplicable` and `noneOfTheAbove` options can't be unselected

marta- commented:
> The questionnaire editor has the ability to add descriptions to question matrix answer, but they are not displayed on the form editor

Addressed in cdbfe5b.

> With multi select checkboxes, `none` and `notApplicable` checkboxes cannot be manually deselected

I can't believe we missed this. Fixed in 48edd5b.

sashaandjic commented:
A small problem when descriptions are introduced on some, not all options
![image](https://user-images.githubusercontent.com/23243244/168653029-7b328909-d978-44c5-b2dc-662827e5ef6a.png)

marta- committed:
CARDS-1400: New Questionnaire entry type: Question matrix
Small layout fix for when only some of the answer options of a question matrix have descriptions

marta- commented:
> A small problem when descriptions are introduced on some, not all options

Addressed in 9a9d172 :
![image](https://user-images.githubusercontent.com/651980/168656354-779aa845-ce6d-46dd-ac22-f08106d814ad.png)

![image](https://user-images.githubusercontent.com/651980/168656422-05b49235-97dd-4257-927f-ad6b21a3ddca.png)

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/168661512-3ebbdc97-314c-4daf-9240-5097c42e1982.png)

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/internal/QuestionMatrixEditor.java:
```diff
@@ -0,0 +1,263 @@
+        }
+    }
+
+    // When something changes in a node deep in the content tree, the editor is invoked starting with the root node,
+    // descending to the actually changed node through subsequent calls to childNodeChanged. The default behavior of
+    // DefaultEditor is to stop at the root, so we must override the following two methods in order for the editor to be
+    // invoked on non-root nodes.
```
This is not very accurate in this class, we actually do useful things in this method, not just forwarding the edit notification down the tree.

sdumitriu committed:
[misc] Comment improvements
